### PR TITLE
Fix various aspects of pin to specific medium

### PR DESCRIPTION
**Pull request by yuzhu**
1. Added more comprehensive testing
2. Correct client side File Info
3. Have the pinning status of a directory propagated to the files

### CODE REVIEW

File: job/common/src/main/java/alluxio/job/plan/replicate/MoveConfig.java [hunk 1]
```diff
@@ -72,7 +72,7 @@ public String getName() {
 
   @Override
   public Collection<String> affectedPaths() {
-    if (StringUtils.isEmpty(mPath)) {
+    if (!StringUtils.isEmpty(mPath)) {
```
Comment (ggezer) [inline job/common/src/main/java/alluxio/job/plan/replicate/MoveConfig.java:75]: were we not moving anything all this time?

File: job/common/src/main/java/alluxio/job/wire/PlanInfo.java [hunk 2]
```diff
@@ -225,13 +226,14 @@ public boolean equals(Object o) {
         && Objects.equal(mResult, that.mResult)
         && Objects.equal(mLastUpdated, that.mLastUpdated)
         && Objects.equal(mName, that.mName)
-        && Objects.equal(mDescription, that.mDescription);
+        && Objects.equal(mDescription, that.mDescription)
+        && Objects.equal(mAffectedPaths, that.mAffectedPaths);
```
Comment (ggezer) [inline job/common/src/main/java/alluxio/job/wire/PlanInfo.java:230]: Is this coming from `MoveConfig`? Will it be sorted always?

**Review by ggezer — COMMENTED** [review]
looks good overall. just have some questions.

File: job/common/src/main/java/alluxio/job/plan/replicate/MoveConfig.java (hunk 1 above)
Comment (yuzhu) [reply job/common/src/main/java/alluxio/job/plan/replicate/MoveConfig.java:75]: we weren't displaying them in the affectedPaths

File: job/common/src/main/java/alluxio/job/wire/PlanInfo.java (hunk 2 above)
Comment (yuzhu) [reply job/common/src/main/java/alluxio/job/wire/PlanInfo.java:230]: not necessarily, but we insert them into the list in certain order, so they should stay in that order and this comparison would work. 
In most cases, this list has one element only.

File: core/common/src/main/java/alluxio/grpc/GrpcUtils.java [hunk 3]
```diff
@@ -234,6 +235,7 @@ public static FileInfo fromProto(alluxio.grpc.FileInfo pInfo) {
         .setBlockSizeBytes(pInfo.getBlockSizeBytes()).setCreationTimeMs(pInfo.getCreationTimeMs())
         .setCompleted(pInfo.getCompleted()).setFolder(pInfo.getFolder())
         .setPinned(pInfo.getPinned()).setCacheable(pInfo.getCacheable())
+        .setMediumTypes(new HashSet<>(pInfo.getMediumTypeList()))
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/common/src/main/java/alluxio/grpc/GrpcUtils.java:238]: yes, and changed to immutable type.

**Comment by yuzhu** [comment]
@ggezer PTAL

**Review by ggezer — APPROVED** [review]

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.6 please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.5 please

**Comment by yuzhu** [comment]
alluxio-bot, cherry-pick this to branch-2.6 please

**Pull request merged**